[FEATURE] Afficher les informations CPF dans PixAdmin (PIX-2838)

P-Jeremy opened the pull request:
## :unicorn: Problème
On souhaite afficher les données CPF dans la section candidat de la page d'informations d'une certification.

## :robot: Solution
Ajouter les données manquantes (sexe, pays de naissance, code postal de naissance, code Insee de naissance)

## :rainbow: Remarques
Affichage sous feature toggle FT_IS_NEW_CPF_DATA_ENABLED

Les données CPF concernant le genre et le pays de naissance sont remplacées par un "Non renseigné" si absentes, les autres sont remplacées par un ' - '.

Ajout d'un quick fix sur l'affichage des bouttons de la barre d'action:

![image](https://user-images.githubusercontent.com/3769147/126802481-371c044e-5b1a-4422-812a-d6b053f92113.png)


TODO
- [x] Ajouter les infos CPF côté SCO ( voir [ici](https://github.com/1024pix/pix/pull/3213#discussion_r673807283) )

## :100: Pour tester
- Lancer pix-api avec FT_IS_NEW_CPF_DATA_ENABLED à true

NON SCO :

- créer une session de certification
- ajouter des candidats (Un né en France avec code postal, un né en France avec code Insee et un né à l'étranger)
- rejoindre la session de certification avec les 3 candidats
- aller voir les détails de ces certification dans pix-admin
- constater l'affichage des données CPF

![image](https://user-images.githubusercontent.com/37305474/126338065-b97fcf6c-5028-4e3d-bb6b-795ecf2a430c.png)

SCO:

- Aller à la session 1
- ajouter les candidats Blue Ivy Carter et George de Cambridge 
- rejoindre la session de certification avec les 2 candidats (george.cambridge@example.net, blueivy.carter@example.net)
- aller voir les détails de ces certification dans pix-admin
- constater l'affichage des données CPF

![image](https://user-images.githubusercontent.com/37305474/126624258-52b14292-33a1-45e7-961a-2d9649f8d90b.png)

![image](https://user-images.githubusercontent.com/37305474/126624300-0ab17671-432a-4ce7-b28d-5503834a3c87.png)

P-Jeremy committed:
feat(api): Add candidate cpf information to certification

P-Jeremy committed:
feat(admin): Add CPF feature toggle

P-Jeremy committed:
feat(admin): Add cpf information to Certification model

P-Jeremy committed:
feat(admin): Add shouldDisplayCPFInformation getter

laura-bergoens commented on api/lib/infrastructure/repositories/general-certification-information-repository.js:
```diff
@@ -16,12 +18,55 @@ module.exports = {
       const certificationIssueReportsDTO = await knex('certification-issue-reports')
         .where({ certificationCourseId });
 
+      if (certificationCourseDTO.birthPostalCode) {
```
je peux pas mettre de commentaire plus haut, mais le orderBy de la première requête ne sert absolument à rien

P-Jeremy replied:
Il est introduit dans une PR précédente, je l'enlève ?

P-Jeremy commented on admin/tests/unit/controllers/authenticated/certifications/certification/informations_test.js:
```diff
@@ -163,6 +163,32 @@ module('Unit | Controller | authenticated/certifications/certification/informati
+          isNewCpfDataEnabled: true,
+        },
+      });
+
```
@laura-bergoens ~~Par rapport à cette [conversation](https://github.com/1024pix/pix/pull/3212#discussion_r672971754), C'est pas mieux de faire comme ça pour les test unitaires ? Ici le service n'est pas écrasé par autre chose, on stub juste sa valeur de retour ?~~

EDIT: This is the way https://guides.emberjs.com/release/testing/testing-components/#toc_stubbing-services

P-Jeremy committed:
feat(admin): Display cpf information

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr3213.review.pix.fr
- App (.org): https://app-pr3213.review.pix.org
- Orga: https://orga-pr3213.review.pix.fr
- Certif: https://certif-pr3213.review.pix.fr
- Admin: https://admin-pr3213.review.pix.fr
- API: https://api-pr3213.review.pix.fr/api/

 Please check it out!

P-Jeremy committed:
feat(api): Remove unecessary orderBy

MathieuGilet commented on api/lib/infrastructure/repositories/general-certification-information-repository.js:
```diff
@@ -16,12 +18,55 @@ module.exports = {
       const certificationIssueReportsDTO = await knex('certification-issue-reports')
         .where({ certificationCourseId });
 
+      if (certificationCourseDTO.birthPostalCode) {
+        const candidateBirthCity = await knex('certification-cpf-cities')
+          .select('name', 'postalCode')
+          .where({ postalCode: certificationCourseDTO.birthPostalCode })
+          .first();
+
+        certificationCourseDTO.birthplace = candidateBirthCity.name;
+        certificationCourseDTO.birthPostalCode = candidateBirthCity.postalCode;
+        certificationCourseDTO.birthCountry = 'FRANCE';
+      }
+
+      if (certificationCourseDTO.birthINSEECode) {
+
+        if (_INSEECodeIsForeign(certificationCourseDTO.birthINSEECode)) {
+          const candidateBirthCountry = await knex('certification-cpf-countries')
+            .select('commonName')
+            .where({ code: certificationCourseDTO.birthINSEECode })
+            .first();
+
+          certificationCourseDTO.birthCountry = candidateBirthCountry.commonName;
+        }
+
+        if (! _INSEECodeIsForeign(certificationCourseDTO.birthINSEECode)) {
+
+          const candidateBirthCity = await knex('certification-cpf-cities')
+            .select('name', 'INSEECode')
+            .where({ INSEECode: certificationCourseDTO.birthINSEECode })
+            .first();
+
+          certificationCourseDTO.birthplace = candidateBirthCity.name;
+          certificationCourseDTO.birthINSEECode = candidateBirthCity.INSEECode;
+          certificationCourseDTO.birthCountry = 'FRANCE';
+        }
+      }
+
```
Pourquoi ne pas avoir reporté aussi les info de ville et pays (peut-être que pays vu qu'il y a `birthplace`) depuis le candidat plutôt que de devoir rechercher ces valeurs à chaque fois ?  @laura-bergoens  ?

P-Jeremy replied:
Dans le certif course, birthPlace c'est la ville, par contre le pays n'existe pas dans le certif course

P-Jeremy replied:
@MathieuGilet Je ne suis pas sûr que la Laura que tu as tagué puisse nous aider :rofl:

MathieuGilet replied:
Je pense que si, c'est elle qui avait rajouté les colonnes INSEECode et postalCode dans le certification course (https://1024pix.atlassian.net/browse/PIX-2716).

P-Jeremy replied:
Ce que je veux dire,check la personne que tu as tagué :wink:, ce n'est pas @laura-bergoens

P-Jeremy replied:
Après pour ce bout de code, c'est probablement overkill d'aller checker en base la ville à partir du code insee. On peut juste reporter ce qu'il y'a dans la table. Par contre le pays, pas le choix

MathieuGilet replied:
Ah, bien vu ;)

MathieuGilet commented on api/lib/infrastructure/repositories/general-certification-information-repository.js:
```diff
@@ -4,24 +4,68 @@ const { NotFoundError } = require('../../domain/errors');
 };
 
+function _INSEECodeIsForeign(INSEECode) {
+  return INSEECode !== FRANCE_INSEE_CODE && _INSECODEStartsWith99(INSEECode);
```
```suggestion
  return INSEECode !== FRANCE_INSEE_CODE;
```
C'est suffisant ici pour savoir si c'est un pays étranger ou non

P-Jeremy replied:
Nein, si le code insee est 75115, ce n'est pas un pays étranger :)

MathieuGilet replied:
Arf oui c'est pas faux. Du coup, dans la méthode en dessous, un `INSEECode.startsWith('99')` devrait être mieux que des `charAt`.

P-Jeremy committed:
feat(api): Add birthCountry in certif course migration

laura-bergoens commented on admin/app/controllers/authenticated/certifications/certification/informations.js:
```diff
@@ -67,6 +68,10 @@ export default class CertificationInformationsController extends Controller {
     return this.certification.status === 'cancelled';
   }
 
+  get isNewCpfDataToggleEnabled() {
```
j'orienterai le nom de ce getter vers ce qu'il autorise plutot que de l'implém :
shouldDisplayCPFInformation ?

laura-bergoens commented on api/db/migrations/20210721084659_add-birthCountry-in-certification-course-table.js:
```diff
@@ -0,0 +1,12 @@
+
+exports.up = async function(knex) {
+  await knex.schema.table('certification-courses', (table) => {
+    table.string('birthCountry');
```
c'est pas tout de l'ajouter dans la migration, il faut donc aussi penser à recopier la donnée vers le certification course lorsqu'on lance un test de certification pour être ISO partout.
A faire ici -> api/lib/domain/models/CertificationCourse.js:55

laura-bergoens replied:
Et puis je pense qu'il y aurait quelque chose à faire ici aussi pour la prescription de certification SCO -> api/lib/domain/usecases/enroll-students-to-session.js:30

P-Jeremy replied:
Attends attends, ça vient dans un prochain commit :sweat_smile:

P-Jeremy replied:
@laura-bergoens En fait pour les schooling registrations, on ne gère que les candidats Français ? 

Dans le `enroll-candidate` on fait 
```const scoCertificationCandidates = students.map((student) => {
    return new SCOCertificationCandidate({
      firstName: student.firstName.trim(),
      lastName: student.lastName.trim(),
      birthdate: student.birthdate,
      birthINSEECode: student.birthCityCode,
      sex: student.sex,
      sessionId,
      schoolingRegistrationId: student.id,
    });
  });
```
  
  Du coup le birthInseeCode de ce type de candidate correspond toujours à une birthCity, donc le candidat est toujours née en France, correct ?

P-Jeremy replied:
La méthode static du CertificationCourse

```
static from({ certificationCandidate, challenges, verificationCode, maxReachableLevelOnCertificationDate }) {
    return new CertificationCourse({
      userId: certificationCandidate.userId,
      sessionId: certificationCandidate.sessionId,
      firstName: certificationCandidate.firstName,
      lastName: certificationCandidate.lastName,
      birthdate: certificationCandidate.birthdate,
      birthPostalCode: certificationCandidate.birthPostalCode,
      birthINSEECode: certificationCandidate.birthINSEECode,
      birthCountry: certificationCandidate.birthCountry,
      sex: certificationCandidate.sex,
      birthplace: certificationCandidate.birthCity,
      externalId: certificationCandidate.externalId,
      isV2Certification: true,
      challenges,
      verificationCode,
      maxReachableLevelOnCertificationDate,
    });
  }
 ```
 
 prend un certif candidat en paramètre. Celui ci voit ses infos CPF misent à jour lors de l'import via l'ods
 
 ```
   if (version === '1.5') {
    const cpfBirthInformation = await certificationCpfService.getBirthInformation({
      ...certificationCandidate,
      certificationCpfCityRepository,
      certificationCpfCountryRepository,
    });

    if (cpfBirthInformation.hasFailed()) {
      throw new CpfBirthInformationValidationError(cpfBirthInformation.message);
    }

    certificationCandidate.updateBirthInformation(cpfBirthInformation);
  }
 ```
 
 Mais les candidats sco ne passent jamais par le certificationCpfService, du coup ce type de candidat n'a qu'un INSEECode, jamais de birthCity ou de birthCountry

P-Jeremy replied:
Question: On ajoute le birthCountry forcé à 'FRANCE' dans `SCOCertificationCandidate` et on va chercher la birthCity via l'insee code en partant du principe que `birthINSSECode === student.birthCityCode` ?

Mais cela présume que tous les candidats sco sont nés en France... Hors ce n'est pas forcément le cas :exploding_head:

laura-bergoens commented on api/lib/infrastructure/repositories/general-certification-information-repository.js:
```diff
@@ -4,24 +4,61 @@ const { NotFoundError } = require('../../domain/errors');
         .where({ certificationCourseId });
 
+      if (certificationCourseDTO.birthPostalCode) {
+        const candidateBirthCity = await knex('certification-cpf-cities')
+          .select('name', 'postalCode')
+          .where({ postalCode: certificationCourseDTO.birthPostalCode })
+          .first();
+
+        certificationCourseDTO.birthplace = candidateBirthCity.name;
+        certificationCourseDTO.birthPostalCode = candidateBirthCity.postalCode;
+        certificationCourseDTO.birthCountry = 'FRANCE';
+      }
+
+      if (certificationCourseDTO.birthINSEECode) {
+
+        if (_INSEECodeIsForeign(certificationCourseDTO.birthINSEECode)) {
+          const candidateBirthCountry = await knex('certification-cpf-countries')
+            .select('commonName')
+            .where({ code: certificationCourseDTO.birthINSEECode })
+            .first();
+
+          certificationCourseDTO.birthCountry = candidateBirthCountry.commonName;
+        }
+
+        if (! _INSEECodeIsForeign(certificationCourseDTO.birthINSEECode)) {
+
+          certificationCourseDTO.birthCountry = 'FRANCE';
+        }
+      }
```
je ne comprends pas trop pourquoi y'a besoin de tout ça.
Il me semblait que quand on inscrivait un candidat via un code postal (par exemple), la birthCity du certification-candidate était remplie avec la bonne valeur (et donc c'est présent dans le certif course a priori, pas besoin de retro engineeré les codes INSEE et code postaux, non ?)

P-Jeremy replied:
Pas faux...

P-Jeremy committed:
feat(api): Add birthCountry in certif course

P-Jeremy committed:
feat(api): Add valid country code to seeds

P-Jeremy committed:
feat(api): Add country and city names to sco certif candidate

P-Jeremy committed:
feat(api): Add sco certifiable profiles in seeds

laura-bergoens commented on api/lib/domain/usecases/enroll-students-to-session.js:
```diff
@@ -26,12 +28,19 @@ module.exports = async function enrollStudentsToSession({
-  const scoCertificationCandidates = students.map((student) => {
+  const scoCertificationCandidates = await bluebird.map(students, async (student) => {
+
+    const code = '99' + student.birthCountryCode;
+
+    const studentCountry = await certificationCpfCountryRepository.getByInseeCode({ code });
```
Si on recherche le pays à chaque boucle, on va faire autant de requêtes BDD en plus que de candidats à inscrire. Même si la requête est "gratos" parce que pas compliquée, c'est dommage de solliciter autant le pool de connexions.
Je pense que ça peut le faire d'utiliser le repo country-repository, de faire un findAll en dehors de la boucle, pour disposer de tous les pays à portée de main (450pays en mémoire c'est que dalle), et de rechercher le pays dans cette collection dans la boucle

laura-bergoens replied:
En plus quand on regarde l'implém de la méthode de répo que tu as créé, je lui trouve deux défauts :
- Elle retourne null si elle ne trouve rien (donc craque ça fait une erreur 500 ici)
- Elle choisit arbitrairement un des pays au code donné, sachant qu'il peut y en avoir plusieurs

P-Jeremy committed:
feat(api): Add FT_IS_NEW_CPF_DATA_ENABLED to validate env variable

octo-topi reviewed (approved):
Mob review

laura-bergoens commented on api/lib/domain/usecases/enroll-students-to-session.js:
```diff
@@ -26,12 +28,21 @@ module.exports = async function enrollStudentsToSession({
+
+    const studentInseeCountryCode = INSEE_PREFIX_CODE + student.birthCountryCode;
+
+    const studentCountry = countries.find((country) => country.code === studentInseeCountryCode);
```
Je pèterai quand même une erreur ici si jamais le pays n'est pas trouvé, que je ferai remonter jusqu'à PixCertif en mode "L'élève NOM PRENOM a été inscrit avec un code pays de naissance invalide. Veuillez corriger ses informations sur l'espace PixOrga de l'établissement ou contacter le support Pix" en attendant, puis on fera relire à Anne-so

aceol committed:
tech:(api) Throw error when birth country is not found

aceol committed:
tech(certif): Move notification component to be available globally in certif app

aceol committed:
feat(certif): Display specific message for 422 errors

If there is a 422 error (which may be raised by an unknown country) if want to display the specific error message.

aceol committed:
fix(certif): Fix action bar buttons display

aceol committed:
fixup! feat(certif): Display specific message for 422 errors
